File: cpp/day2.cpp
/*
 * Solve day 2.
 *
 * On the right are notes to myself about cpp, revealing what I do and don't
 * know about the language.
 */

#include "utils.h"


/*
 * digest_line --- return a count of each letter appearing in input word
 */
std::unordered_map<char, int>
digest_line(std::string word)
{
  std::unordered_map<char, int> seen;
  for (char c : word)
  {
    seen[c] += 1;
  }
  return seen;
}


/*
 * contains_value --- returns True if key is in the unordered_map
 */
bool contains_value(std::unordered_map<char, int> map, int rep)
{
  for (auto iter = map.begin(); iter != map.end(); ++iter)
  {
    if (iter->second == rep)    // the map is a bunch of std::pair<char, int>
    {                           // Thus to access the int, one uses second to
      return true;              // get the int part. This is the same as
    }                           // (*iter).second.
  }
  return false;
}



void part1(std::string content)
{
  int doubles {0};
  int triples {0};
  std::string line;
  std::istringstream input(content);

  while (std::getline(input, line))
  {
    std::unordered_map<char, int> seen = digest_line(line);
    if (contains_value(seen, 2)) { doubles += 1; }
    if (contains_value(seen, 3)) { triples += 1; }
  }
  std::cout << doubles * triples << std::endl;
}


/*
 * distance --- compute naive Hamming distance between word1 and word2
 */
int distance(std::string word1, std::string word2)
{
  int ndiffs {0};
  for (uint i = 0; i < word1.length(); ++i)
  {
    if (word1[i] != word2[i]) { ndiffs += 1; }
  }
  return ndiffs;
}


/*
 * common_substring --- return common substring between word1 and word2
 */
std::string common_substring(std::string word1, std::string word2)
{
  std::string res = "";
  for (uint i = 0; i < word1.length(); ++i)
  {
    if (word1[i] == word2[i]) { res += word1[i]; }
  }
  return res;
}


void part2(std::string content)
{
  std::string line;
  std::istringstream input(content);
  while (std::getline(input, line))
  {
    std::string line2;
    std::istringstream input2(content);

    while (std::getline(input2, line2))
    {
      if (distance(line, line2) == 1)
      {
        std::cout << common_substring(line, line2) << std::endl;
        return;
      }
    }
  }
}


int main()
{
  std::string content;
  content = read_file("inputs/input2.txt");
  part1(content);
  part2(content);
}

File: python/day2.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

"""
Solve day 2.
"""

from collections import Counter
from utils import input_lines


def digest_line(line):
    """
    Parses a line into a dictionary containing letter counts.
    """
    seen = Counter(line.strip())
    return seen


def contains_nple(seen_dict, reps=2):
    """
    Returns True if line contains a letter exactly `reps` times.
    """
    if reps in seen_dict.values():
        return True
    return False


def do_part_1():
    """
    Solves part 1
    """
    digested_lines = list(map(digest_line, input_lines(2)))
    # Poor man's partial
    doubles = sum(map(lambda l: contains_nple(l, reps=2), digested_lines))
    triples = sum(map(lambda l: contains_nple(l, reps=3), digested_lines))
    print(doubles * triples)
    return doubles * triples


def distance(word1, word2):
    """
    Returns the number of different letters between word1 and word2
    """
    ndiffs = 0
    for a, b in zip(word1, word2):
        if a != b:
            ndiffs += 1
    return ndiffs


def common_substring(word1, word2):
    """
    Returns the common substring between word1 and word2
    """
    ret = ""
    for a, b in zip(word1, word2):
        if a == b:
            ret += a
    return ret


def do_part_2():
    """
    Solves part 2
    """
    lines = input_lines(2)
    for line1 in lines:
        for line2 in lines:
            if distance(line1, line2) == 1:
                common = common_substring(line1, line2)
                print(common)
                return common
    return ""


if __name__ == "__main__":
    do_part_1()
    do_part_2()
